### PR TITLE
Avoid casting input to _fft_helper.

### PR DESCRIPTION
**Pull request by bartvm**
See #11374.

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks for the PR - I'm happy to merge a change of this sort, but we need to ensure it's compatible with strict dtype promotion mode, which means that we can't have internal code generate operations between arrays of differing dtypes.

File: jax/_src/scipy/signal.py [hunk 1]
```diff
@@ -180,6 +180,7 @@ def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft, sides):
   result = detrend_func(result)
 
   # Apply window by multiplication
+  result = result.astype(win.dtype)
```
[the remark this answers is not in the document]
Comment (bartvm) [reply jax/_src/scipy/signal.py]: What about the case where a real input and complex window are passed? This happens in the tests (e.g., `scipy_signal_test.LaxBackedScipySignalTests testMethod=testWelchWithDefaultStepArgsAgainstNumpy_shape=float32[2,3,389,5]_nperseg=17_noverlap=13_use_nperseg=False_use_overlap=False_axis=2`).

For computational efficiency I think you might in fact want to allow the signal and window to be different types. In this particular case, the generation of the input frames can be done in the reals (much faster) while still allowing the use of a complex window.

**Comment by bartvm** [comment]
Here's a new proposal which passes the tests:

* The input is kept as is when passed to `_fft_helper` so that the frame generation can be done in the original dtype
* The window is generated with the same dtype as the input, so that windowing can be done in that dtype too
  * But the calling function might have passed a complex window, in this case the frames are cast to complex just before the windowing
* The output of the FFT is always complex, so the scale needs to be complex as well

**Review by jakevdp — APPROVED** [review]
Looks good, thanks!

**Pull request merged**